components/script/dom/webgl2renderingcontext.rs

```diff
@@ -147,6 +156,196 @@ impl WebGL2RenderingContext {
             slot.set(None);
         }
     }
+
+    fn calc_read_pixel_formats(
+        &self,
+        pixel_type: u32,
+        format: u32,
+    ) -> Result<(&[Type], usize), WebGLError> {
+        let allowed_array_types = match pixel_type {
+            constants::BYTE => &[Type::Int8][..],
+            constants::SHORT => &[Type::Int16][..],
+            constants::INT => &[Type::Int32][..],
+            constants::UNSIGNED_BYTE => &[Type::Uint8, Type::Uint8Clamped][..],
+            constants::UNSIGNED_SHORT |
+            constants::UNSIGNED_SHORT_4_4_4_4 |
+            constants::UNSIGNED_SHORT_5_5_5_1 |
+            constants::UNSIGNED_SHORT_5_6_5 => &[Type::Uint16][..],
+            constants::UNSIGNED_INT |
+            constants::UNSIGNED_INT_2_10_10_10_REV |
+            constants::UNSIGNED_INT_10F_11F_11F_REV |
+            constants::UNSIGNED_INT_5_9_9_9_REV => &[Type::Uint32][..],
+            constants::FLOAT => &[Type::Float32][..],
+            constants::HALF_FLOAT => &[Type::Uint16][..],
+            _ => return Err(InvalidEnum),
+        };
+        let channels = match format {
+            constants::ALPHA | constants::RED | constants::RED_INTEGER => 1,
+            constants::RG | constants::RG_INTEGER => 2,
+            constants::RGB | constants::RGB_INTEGER => 3,
+            constants::RGBA | constants::RGBA_INTEGER => 4,
+            _ => return Err(InvalidEnum),
+        };
+        Ok((allowed_array_types, channels))
+    }
+
+    fn calc_read_pixel_sizes(
+        &self,
+        width: i32,
+        height: i32,
+        bytes_per_pixel: usize,
+    ) -> Result<(usize, usize, usize), WebGLError> {
+        if width < 0 || height < 0 {
+            return Err(InvalidValue);
+        }
+
+        let pixels_per_row = if self.texture_pack_row_length.get() > 0 {
+            self.texture_pack_row_length.get()
+        } else {
+            width as usize
+        };
```

Comment:
I don't see any spec link in the code, which would be helpful to understand this kind of snippet.